e2e/rapid_test.go

```diff
@@ -0,0 +1,598 @@
+	slice []bool
+	mtx   sync.RWMutex
+}
+
+func (bs *BoolSlice) Set(i int, b bool) {
+	bs.mtx.Lock()
+	defer bs.mtx.Unlock()
+	bs.slice[i] = b
+}
+func (bs *BoolSlice) Get(i int) bool {
+	bs.mtx.Lock()
+	defer bs.mtx.Unlock()
+	return bs.slice[i]
+}
+
+func (bs *BoolSlice) Iterate(f func(k int, v bool)) {
+	bs.mtx.RUnlock()
+	defer bs.mtx.RUnlock()
+	for k, v := range bs.slice {
+		f(k, v)
+	}
+}
+
+func (bs *BoolSlice) CalculateNum(val bool) int {
+	bs.mtx.RLock()
+	defer bs.mtx.RUnlock()
+	nm := 0
+	for _, v := range bs.slice {
+		if v == val {
+			nm++
+		}
+	}
+	return nm
+}
+
+func (bs *BoolSlice) String() string {
+	bs.mtx.RLock()
+	defer bs.mtx.RUnlock()
+	return fmt.Sprintf("%v", bs.slice)
+}
+
+func generateNode(id int, transport *fakeTransport) (*pbft.Pbft, chan time.Time) {
+	node := pbft.New(key(strconv.Itoa(id)), transport,
+		pbft.WithTracer(trace.NewNoopTracerProvider().Tracer("")),
+		func(config *pbft.Config) {},
+		pbft.WithLogger(log.New(io.Discard, "", 0)),
+	)
+
+	timeoutChan := make(chan time.Time)
+	// round timeout mock
+	node.SetRoundTimeoutFunction(func() <-chan time.Time {
+		return timeoutChan
+	})
+
+	transport.nodes = append(transport.nodes, node)
+	return node, timeoutChan
+}
+
+func generateCluster(numOfNodes int, transport *fakeTransport) ([]*pbft.Pbft, []chan time.Time) {
+	nodes := make([]string, numOfNodes)
+	timeoutsChan := make([]chan time.Time, numOfNodes)
+	cluster := make([]*pbft.Pbft, numOfNodes)
+	for i := 0; i < numOfNodes; i++ {
+		cluster[i], timeoutsChan[i] = generateNode(i, transport)
+		nodes[i] = strconv.Itoa(i)
+	}
+
+	for i, node := range cluster {
+		i := i
+		node := node
+
+		valsAsNode := []pbft.NodeID{}
+		for _, i := range nodes {
+			valsAsNode = append(valsAsNode, pbft.NodeID(i))
+		}
+		vv := valString{
+			nodes: valsAsNode,
+		}
+
+		node.SetBackend(&BackendFake{
+			nodes:  nodes,
+			nodeId: i,
+			IsStuckMock: func(num uint64) (uint64, bool) {
+				return 0, false
+			},
+			ValidatorSetList: &vv,
+		})
+	}
+	return cluster, timeoutsChan
+}
+
+func runClusterCycle(cluster []*pbft.Pbft, callNumber int, stuckList, doneList *BoolSlice) error {
+	wg := errgroup.Group{}
+	for i := range cluster {
+		i := i
+		state := cluster[i].GetState()
+		isLocked := cluster[i].IsLocked()
+		wg.Go(func() (err1 error) {
+			wgTime := time.Now()
+			exitCh := make(chan struct{})
+			deadlineTimeout := waitDuration
+			deadline := time.After(deadlineTimeout)
+			if stuckList.Get(i) {
+				return nil
+			}
+
+			if doneList.Get(i) {
+				return nil
+			}
+
+			go func() {
+				stuckList.Set(i, true)
+				defer func() {
+					stuckList.Set(i, false)
+				}()
+				cluster[i].RunCycle(context.Background())
+				close(exitCh)
+			}()
+			select {
+			case <-exitCh:
+			case <-deadline:
+			}
+
+			//useful for debug
+			_, _, _ = state, wgTime, isLocked
+			//if time.Since(wgTime) > waitDuration {
+			//	fmt.Println("wgitme ", state, i, callNumber, time.Since(wgTime), err1, isLocked)
+			//}
+			//
```

Comment:
Is this part necessary to leave or is it a leftover?

Reply:
I'd prefer to save it for debug purposes.